kimchi/cairo/src/turshi/constraint.rs

```diff
@@ -0,0 +1,519 @@
+                0 => res = op1,        // right part is single operand
+                1 => res = op0 + op1,  // right part is addition
+                2 => res = op0 * op1,  // right part is multiplication
+                _ => unimplemented!(), // invalid instruction
+            }
+        } else {
+            // multiple bits take value 1
+            unimplemented!(); // invalid instruction
+        }
+
+        // Compute new value of pc
+        match pc_up {
+            0 => {
+                // next instruction is right after the current one
+                next_pc = pc + size // the common case
+            }
+            1 => {
+                // next instruction is in res
+                next_pc = res // absolute jump
+            }
+            2 => {
+                // relative jump
+                next_pc = pc + res // go to some address relative to pc
+            }
+            4 => {
+                // conditional relative jump (jnz)
+                if dst == 0 {
+                    next_pc = pc + size; // if condition false, common case
+                } else {
+                    // if condition true, relative jump with second operand
+                    next_pc = pc + op1;
+                }
+            }
+            _ => unimplemented!(), // invalid instruction
+        }
+
+        // Compute new value of ap and fp based on the opcode
+        if opcode == 1 {
+            // "call" instruction
+            assert_eq!(dst, fp); // checks [ap] contains fp
+            assert_eq!(op0, pc + size); // checks [ap+1] contains instruction after call
+
+            // Update fp
+            next_fp = ap + 2; // pointer for next frame is after current fp and instruction after call
+
+            // Update ap
+            match ap_up {
+                0 => next_ap = ap + 2, // two words were written so advance 2 positions
+                _ => unimplemented!(), // ap increments not allowed in call instructions
+            }
+        } else if opcode == 0 || opcode == 2 || opcode == 4 {
+            // rest of types of instruction
+            // jumps and increments || return || assert equal
+            match ap_up {
+                0 => next_ap = ap,       // no modification on ap
+                1 => next_ap = ap + res, // ap += <op>
+                2 => next_ap = ap + 1,   // ap++
+                _ => unimplemented!(),   // invalid instruction
+            }
+            match opcode {
+                0 => next_fp = fp,  // no modification on fp
+                2 => next_fp = dst, // ret sets fp to previous fp that was in [ap-2]
+                4 => {
+                    assert_eq!(res, dst); // assert equal result and destination
+                    next_fp = fp // no modification on fp
+                }
+            }
+        } else {
+            unimplemented!(); // invalid instruction
+        }
+
+        // convert to Fp all of it
+
+        constraint();
+    }
+}
+*/
```

Comment:
Oh, that file is indeed important. It will contain (and is containing right now in a more advanced version of that file) the actual Kimchi constraints representing the cairo gate (what we called turshi). But you're right that in this PR we should focus only on the runner and leave all of the other contents for a different PR. The reason that this file was commented out was to ignore it, but I can see this is not the cleanest way to proceed: better have different PRs for each large module.